src/content/components/NodeIcon.js

```diff
@@ -0,0 +1,63 @@
+import React, { PureComponent, PropTypes } from 'react';
+
+const icons = new Map();
+
+function getIconForNode(node) {
+  if (!node.icon) {
+    return Promise.resolve(null);
+  }
+
+  if (icons.has(node.icon)) {
+    return icons.get(node.icon);
+  }
+
+  const result = new Promise(resolve => {
+    const image = new Image();
+    image.src = node.icon;
+    image.referrerPolicy = 'no-referrer';
+    image.onload = () => {
+      resolve(node.icon);
+    };
+    image.onerror = () => {
+      resolve(null);
+    };
+  });
+
+  icons.set(node.icon, result);
+  return result;
+}
+
+class NodeIcon extends PureComponent {
+  constructor(props) {
+    super(props);
+
+    this.state = {
+      className: null,
+    };
+    this._updateState(props);
```

Comment:
I'm not sure how to handle the asynchronicity in the constructor. I think we can't call setState if the component is not mounted. I've seen warnings in the console but I don't see them anymore now.